### PR TITLE
GH-39126: [C++][CI] Fix Valgrind failures

### LINKED ISSUE #39126: [C++][CI] Valgrind errors
### Describe the bug, including details regarding any error messages, version, and platform.

Valgrind errors have been popping up recently:
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=58208&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181

### Component(s)

C++, Continuous Integration

### PR DESCRIPTION
**Pull request by pitrou**
### Are there any user-facing changes?

* Closes: #39126

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@ursabot please benchmark

**Review by bkietz — CHANGES_REQUESTED** [review]
LGTM, just one nit:

File: cpp/src/arrow/array/builder_binary.cc [hunk 1]
```diff
@@ -80,10 +80,11 @@ Status BinaryViewBuilder::AppendArraySlice(const ArraySpan& array, int64_t offse
...
   ARROW_ASSIGN_OR_RAISE(auto data, data_builder_.FinishWithLength(length_));
-  BufferVector buffers = {null_bitmap, data};
-  for (auto&& buffer : data_heap_builder_.Finish()) {
-    buffers.push_back(std::move(buffer));
-  }
+  ARROW_ASSIGN_OR_RAISE(auto byte_buffers, data_heap_builder_.Finish());
+  BufferVector buffers(byte_buffers.size() + 2);
+  buffers[0] = std::move(null_bitmap);
+  buffers[1] = std::move(data);
+  std::move(byte_buffers.begin(), byte_buffers.end(), byte_buffers.begin() + 2);
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_binary.cc:87]:
```suggestion
  std::move(byte_buffers.begin(), byte_buffers.end(), buffers.begin() + 2);
```

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Pull request merged**